Submit Step3 : 스프링 부트로 QA 게시판 구현 1

Hyune-c opened the pull request:
### # Heroku
https://codesquad-qnaboard.herokuapp.com/

### # 기능 구현
- Answer 을 제외하고 구현 ... complete
- 에러 핸들링에 대한 구현
  - CustomError ... complete
  - DefaultError ... working

### # Refactoring 예정
- Error 핸들링에 대한 개선 
  - Enum Stirng 비교가 아닌 더 좋은 방법을 고민 중입니다.

### # 자료 정리
[Error Hanling](https://github.com/Hyune-c/TIL/blob/master/Spring/Error%20Handling.md)
### # 질문 
1. Error 처리 방법이 적절한지 질문 드립니다.
  - Enum 의 사용이 적절한가요?
  - 에러 처리가 일부만 되고 있습니다.
    - @ControllerAdvice 를 통해 제가 만든 예외 (CustomException) 는 잘 처리가 되었습니다.
    - 하지만 그 외 에러들은 모두 Exception.class 를 만들어 주었음에도 찾지 못하고 404 가 떨어지고 있습니다.
  - @RequestMapping 는 적절히 사용해보았지만 @ResponseStatus 는 사용에 실패하였습니다.
(적용을 몇번 시도해보다 실패해서 지금의 형태로 일단 만들었습니다 ㅠ)
2. Optional 의 orElseThrow 를 사용해보았습니다.   
  적절한 사용법일까요?
3. MVC 모델에 따르면 DB 제어를 포함한 로직 처리는 모두 model 에서 하도록 되어 있습니다.
하지만 제 코드는 DB 처리도 controller 안에서 하는 등 model 을 명시적으로 나누지 않은 듯 합니다.
이 부분을 개선하려면 어떻게 만들어야 될까요?

4. 글 작성 날짜의 formatting 에 대해 좀 더 잘 처리해보고 싶습니다.
- 목표 
  - bean, DB  : LocalDateTime, .hbs : 지정된 format
  - get 메소드의 조작 없이
- 시도해본 방법
- @DateTimeFormat 
    - 컴파일 에러는 발생하지 않았지만 의도한 내용이 적용되지 않았습니다.
    - mvc configuration 을 해주어야 @DateTimeFormat 이 작동하는 걸까요?   
(configuration 을 시도하다 실패했습니다 ㅠ)    
- 아래의 방법은 json 처리를 위한 직렬화 작업이라고 나오던데, 위의 내용과는 상관없는 접근일까요?
  

```java
  // 원래 소스
  private LocalDateTime createdDateTime;

  public String getCreatedDateTime() {
    DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
    return createdDateTime.format(dateTimeFormatter);
  }
```
```java
  // 의도한 소스
  @DateTimeFormat(iso = DateTimeFormat.ISO.NONE)
  private LocalDateTime createdDateTime;
```
```java
  // 검색하여 찾은 get() 메소드 formatting 의 다른 방법
  @JsonFormat(shape=JsonFormat.Shape.STRING, pattern="dd/MM/yyyy")
  public LocalDateTime getCreatedDateTime() {
    return createdDateTime;
  }
```

Hyune-c committed:
refactor : 코드 정리

Hyune-c committed:
feat : lombok + Slf4j 설정

lombok + Slf4j 설정

Hyune-c committed:
feat : DB 초기 데이터 설정 & 코드 정리

data.sql 을 활용한 DB 초기 데이터 설정

Hyune-c committed:
feat : login 구현 1

- 동적인 login.html 작성
- session 저장에 대한 메소드 구현

Hyune-c committed:
feat : handelbars 적용

- 모든 .html 을 .hbs 변경

Hyune-c committed:
refactor : logging 개선

highlight 적용

Hyune-c committed:
feat : User 클래스 개선 & 비밀번호 검증 메소드 추가

- 어노테이션을 활용한 User 클래스 개선
- User 에 validatePassword() 추가
  - 클래스 외부에서 get, set 을 사용하지 않기 위해 별도의 메소드 구현

Hyune-c committed:
feat : login 에 따른 기능 추가 구현

- session 유지를 통해 navigation 동적 변경
- login 시에만 '정보수정' 메뉴가 뜨며 이동 가능하게 수정

Hyune-c committed:
feat : logout 구현

logout 구현

Hyune-c committed:
refactor : Question 추가를 위한 준비

- 질문 초기 data 추가
- 어노테이션 활용을 통한 Question 클래스 정리

Hyune-c committed:
refactor : 설정 변경

- DB 초기 설정 및 로깅 레벨 변경

Hyune-c committed:
refactor : .hbs 의 수정

- /include/navigation.hbs
  - 회원가입 : 로그인 되지 않았을 때만 보이도록 수정
- /questions/form.hbs, show.hbs
  - writer 에서 userId 로 변경
- /questions/update.hbs
  - Question update form 추가
- /users/login.hbs
  - 에러 처리에 대한 내용 추가
- error.hbs, welcome.hbs
  - 에러 처리에 따른 문구 출력

Hyune-c committed:
refactor : .hbs 수정

- 코드 정리
- Question update 를 위한 put 추가

Hyune-c committed:
refactor : get, post 처리를 위한 메소드 분리

Hyune-c committed:
feat : 에러 처리 구조에 대한 변경

- CustomErrorCode.java
  - Enum 을 활용한 관리
- CustomErrorController.java
  - 에러 종류 세분화에 따른 역할 추가

Hyune-c committed:
refactor : getUser() method 추가

User 를 가져오면서 Null 체크를 할 수 있도록 개선

Hyune-c committed:
feat : Question Update 기능 추가

Question Update 기능 추가

Hyune-c committed:
feat : sessionUser 를 가져오는 common method 추가

sessionUser 를 가져오는 common method 추가

Hyune-c committed:
refactor : README.md 업데이트

README.md 업데이트

Hyune-c committed:
refactor : Question 작성과 수정에 대한 개선

- 반복 사용되는 메소드를 Utils 로 합쳐 개선
- Question 클래스에 UserId 를 검증하는 메소드를 추가
- javadoc 설명 개선

Hyune-c committed:
refactor : Question 작성과 상세보기에 대한 개선

- 코드 개선
- javadoc 설명 개선

Hyune-c committed:
refactor : long 타입에서 Long 타입으로 개선

- primitive 타입에서 wrapper 타입으로 개선

Hyune-c committed:
refactor : QuestionController 에 대한 개선

- javadoc 개선

Hyune-c committed:
refactor : 코드 개선

- javadoc 개선
- 불필요한 import 삭제

Hyune-c committed:
refactor : 코드 개선

- javadoc 개선

Hyune-c committed:
refactor : 코드 개선

- javadoc 개선
- 형 변환 개선

Hyune-c committed:
refactor : 코드 개선

- Question.java
  - dateTime 에서 createdDateTime 로 변경
  - formatter 를 get() 메소드 안으로 이동

Hyune-c committed:
refactor : 코드 정리

indent 관리

Hyune-c committed:
refactor : 코드 정리

- 사용되지 않는 파일 삭제

wheejuni commented on src/main/java/com/codessquad/qna/commons/CustomErrorCode.java:
```diff
@@ -0,0 +1,22 @@
+
+  QUESTION_NOT_EXIST("존재하지 않는 Question id 입니다."),
+
+  BAD_REQUEST(HttpStatus.BAD_REQUEST.getReasonPhrase());
```
위 케이스들은 다 문자열을 하드코딩했는데 여기선 다른 상수를 불러왔네요. 특별한 이유가 있나요.

Hyune-c replied:
@wheejuni 하드코딩한 것들은 커스텀 익셉션을 위함이고, 아래의 것은 표준 정의 익셉션을 가져오기 위한 의도였습니다.

wheejuni commented on src/main/java/com/codessquad/qna/commons/CustomErrorController.java:
```diff
@@ -1,35 +1,95 @@
+   * Desc :
+   * Return : /error
+   */
+  @ExceptionHandler(Exception.class)
```
이렇게 선언하면 아무런 예외도 처리할 수가 없습니다. 
생각해보면 아래 `ForbiddenException` 이나 `UserException` 도 모두 다 `Exception` 의 서브클래스에요. 
왜냐면 `RuntimeException` 이 `Exception` 의 서브클래스이고 지금 작성하신 커스텀 에외들이 모두 `RuntimeException` 을 상속했으니까요. https://docs.oracle.com/javase/7/docs/api/java/lang/RuntimeException.html

하위 클래스에 대한 `@ExceptionHandler` 가 있다면 상위 클래스 예외를 catch하는 exception resolver는 무시되므로(별도의 order 설정이 없는 이상 말이죠) 이미 다양한 알려진 예외에 대해 기본 exception resolver를 제공하는 Spring MVC가 이 어드바이스를 사용하지 않은 것은 당연한 것입니다. 

https://stackoverflow.com/questions/38462588/order-of-exceptionhandler 같은 주제로 고민한 사람이 있는 것 같아 링크 가져와봅니다.

wheejuni commented on src/main/java/com/codessquad/qna/commons/CustomErrorController.java:
```diff
@@ -1,35 +1,95 @@
+      model.addAttribute("descMsg", e.getMessage());
+
+      return "/error";
+    } else if (e.getMessage().equals(CustomErrorCode.USER_NOT_EXIST.getMsg())) {
+      model.addAttribute("notExistUser", true);
+    } else if (e.getMessage().equals(CustomErrorCode.USER_NOT_LOGIN.getMsg())) {
+      model.addAttribute("requestLogin", true);
+    } else if (e.getMessage().equals(CustomErrorCode.USER_NOT_MATCHED_PASSWORD.getMsg())) {
+      model.addAttribute("wrongPassword", true);
+    }
```
고민을 많이 하신 것이 티가 나네요. ㅠ
그러나 각 경우에 대해 `UserException` 을 상속하는 다른 커스텀 예외가 있었으면 더 좋았을 뻔 했습니다.

Hyune-c replied:
@wheejuni 
말씀하신게 이해가 잘 안되서요.. 예외 종류를 묶을 수 있는 커스텀 클래스를 만들라는 뜻이신가요? 아직 구현은 못했지만 Enum 을 적용해보려고 시도 했었습니다. (물론 실패............)

wheejuni commented on src/main/java/com/codessquad/qna/commons/CustomErrorController.java:
```diff
@@ -1,35 +1,95 @@
+      model.addAttribute("descMsg", e.getMessage());
+
+      return "/error";
+    } else if (e.getMessage().equals(CustomErrorCode.USER_NOT_MATCHED.getMsg())) {
+      model.addAttribute("notMatchedUser", true);
```
여기두요. 상단 코멘트 참고해주세요

wheejuni commented on src/main/java/com/codessquad/qna/commons/Utils.java:
```diff
@@ -0,0 +1,22 @@
+import javax.servlet.http.HttpSession;
+import java.util.Optional;
+
+public class Utils {
```
클래스 이름 다시 고민해주세요.

wheejuni commented on src/main/java/com/codessquad/qna/question/Question.java:
```diff
@@ -1,69 +1,45 @@
-  private static DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd hh:mm");
-
   @Id
   @GeneratedValue(strategy = GenerationType.IDENTITY)
```
`GenerationType.IDENTITY` 는 어떤 원리로 작동하는 거죠?

Hyune-c replied:
@wheejuni 키 생성을 DB 에 위임하는 것으로 1씩 상승되는 것으로 알고 있습니다.

wheejuni commented on src/main/java/com/codessquad/qna/question/Question.java:
```diff
@@ -1,69 +1,45 @@
-  public String getDateTime() {
-    return dateTime.format(dateTimeFormatter);
+  public boolean validateUserId(User user) {
+    return this.userId.equals(user.getUserId());
```
`User` 클래스에게 동질성의 판단을 위임해야 하는 것은 아니었을지요.

Hyune-c replied:
@wheejuni 맞는거 같아요.. 수정해보겠습니다.

wheejuni commented on src/main/java/com/codessquad/qna/question/QuestionController.java:
```diff
@@ -17,30 +22,109 @@
   @PostMapping("")
-  public String create(Question question) {
+  public String create(Question question, HttpSession session) {
+    Utils.getSessionedUser(session);
```
가져온 결과물을 어디에도 쓰지 않네요. 어떻게 된 거죠?

Hyune-c replied:
@wheejuni 
Desc : Utils.getSessionedUser() 를 통해 login 여부를 검증합니다.
get 메소드 안에 sessionUser 가 null 인 경우 error 처리를 하게하였기에 위에서는 별도의 예외처리를 하지 않았습니다.
하지만 말씀을 듣고보니 메소드명만으로는 의문이 드는 로직일거 같은데.. 메소드 안에서 error 처리를 하더라도 명시적으로 한번 더 처리를 하는게 맞을까요?

honux77 replied:
이름만으로 내용을 알기 어렵네요. 
사소하지만 Utils 는 static import를 하고 checkLoginOrError(session) 뭐 이런 정도의 이름이면 좋았을 것 같습니다.

wheejuni commented on src/main/java/com/codessquad/qna/user/User.java:
```diff
@@ -1,74 +1,35 @@
+import javax.validation.constraints.NotNull;
 
 @Entity
+@Data
```
`@Data` 가 자동생성해주는 메소드는 어떤 것들이 있나요?

Hyune-c replied:
@wheejuni @Getter, @Setter, @ToString, @EqualsAndHashCode, @RequiredArgsConstructor 를 생성해주는 lombok 어노테이션 입니다.

honux77 replied:
가능하면 연습 단계에서는 Lombok을 안 쓰면 좋겠어요. 귀찮더라도 일단은 하나씩 타이핑하는게 좋지 않을까요?

wheejuni commented on src/main/java/com/codessquad/qna/user/User.java:
```diff
@@ -1,74 +1,35 @@
 
 @Entity
+@Data
+@NotNull
```
클래스 레벨에서 `@NotNull` 을 사용하면 어떤 유효성 검증을 하게 되는 건가요?

Hyune-c replied:
@wheejuni 
https://github.com/Hyune-c/TIL/blob/master/Spring/Annotation.md#javaxvalidation-%EB%B9%88-%EA%B2%80%EC%A6%9D
위와 같이 정리하였습니다.

wheejuni reviewed (changes requested):
안녕하세요. 리뷰가 늦어 죄송합니다. 

질문에 대한 답변도 함께 드릴게요.

1 번 질문은 코드상 코멘트로 답변 드렸고, 
`@ResponseStatus` 에 대해서는 이미 보셨을 줄로 알지만 https://springboot.tistory.com/25 링크도 참고해 보세요. 
아무런 Exception Resolver에 걸리지 않은 상태여야 Spring MVC가 제공하는 exception resolver에 pick up되어 처리될 수 있는 애노테이션인 겁니다. 

2. 잘 하셨어요. 
3. `Service` 클래스가 없어서 생기는 문제들인데요, 지금 단계에선 고민 안해도 됩니다. 다만 다음 스텝에서부터는 조금씩 고민해주세요. 
4. https://jojoldu.tistory.com/361 에서 보시면 알 수 있듯이 JSON 기반 요청/응답 본문의 직렬화/역직렬화를 위해 사용하는 애노테이션이구요, 지금 View Template 기반 구현에서는 해당사항이 없는 애노테이션입니다. 
**getter를 사용하는 것이 절대 나쁜 접근이 아니니 그대로 하셔도 됩니다.**

충실한 답변이 되었으면 좋겠습니다! 다른 추가 질문 있으시면 여기에 답글로 남겨주세요. 감사합니다~

Hyune-c committed:
refactor : 공통 메소드 변경

- Utils 에서 CommonUtils 로 변경
- getQuestion() 메소드의 이동 : QuestionController 에서 CommonUtils 로
- validateUserId() 메소드의 이동 : Question 에서 User 로

Hyune-c committed:
refactor : 코드 정리

- 사용되지 않는 코드 삭제

Hyune-c committed:
feat : answer 작성 기능 추가

- answer 관련 java 파일 추가
- 작성 기능 추가

Hyune-c committed:
feat : answer 작성 기능 추가 2

- findByQuestionId() : questionId 매칭되는 Answer 들을 가져옵니다.
- Question 을 보여주는 페이지 (show) 에 answers 를 추가해줍니다.
- 관련된 .hbs 를 수정하고 분리했습니다.

Hyune-c committed:
feat : 코드 정리

- helper 기능을 아직 적용하지 못해 상수로 넣어두었습니다.

Hyune-c committed:
feat : answer 삭제 기능 추가

- delete() 에 관련된 기능을 추가하였습니다.

Hyune-c commented:
@wheejuni 
리뷰를 받은 후 선택 사항을 완료하여 추가 커밋하였습니다.

### # step3 Refactoring
- validateUserId() 메소드의 이동 ... working
- Utils 에서 CommonUtils 로 이름 변경 ... complete 

### # 기능 구현
- Answer 구현하기 실습(선택) 
    - 답변 목록은 모두가 볼 수 있습니다. ... complete
    - 답변 작성은 로그인한 사용자만 가능합니다. ... complete
    - 답변 삭제는 자신의 것만 가능합니다. ... complete

### # 질문
- 여러 곳에서 쓰일 수 있는 메소드를 CommonUtils 로 모으다 보니 DB 에 접근할 때가 애매해졌습니다.  
    - 아래 메소드는 static 이며 Repository 를 매개변수로 전달받게 만들었습니다.
    - Repository 는 Singleton 이기에 전역에서 접근하려 했는데 @autowired 는 static 적용을 할 수가 없었습니다.
    - 어쩔 수 없이 매개변수로 전달하는 형태로 구현했는데 의존성 면에서 좋아보이지 않아 고민됩니다.  

```java
public static Question getQuestion(QuestionRepository questionRepository, Long id) {
```
